File: src/engine/manifest.ts
import { XMLParser } from "fast-xml-parser";
import { Logger } from "../core/logger.js";

export interface DashSegment {
  t?: number;
  d: number;
  r?: number; // Repeat
}

export interface DashStream {
  itag: number;
  mimeType: string;
  codecs: string;
  width?: number;
  height?: number;
  bandwidth: number;
  baseUrl: string;
  startNumber: number;
  timescale: number;
  segments: DashSegment[];
  initialization?: string;
}

/**
 * Result of timestamp-to-segment mapping
 */
export interface SegmentRange {
  /** First segment index to download (0-based from startNumber) */
  startSegment: number;
  /** Last segment index to download (inclusive) */
  endSegment: number;
  /** Offset within the first segment where the start time falls (seconds) */
  trimStartOffset: number;
  /** Total duration to keep from the trimmed start (seconds), or 0 for no limit */
  trimDuration: number;
}

export interface ParsedHlsVariant {
  bandwidth: number;
  width: number;
  height: number;
  codecs: string;
  url: string;
}

export interface ParsedHlsSegment {
  duration: number;
  url: string;
  discontinuity: number;
}

export class ManifestParser {
  static parseDash(xmlContent: string, manifestUrl?: string): DashStream[] {
    const logger = Logger.getInstance();
    const parser = new XMLParser({
      ignoreAttributes: false,
      attributeNamePrefix: "@_",
    });
    const jsonObj = parser.parse(xmlContent);

    const mpd = jsonObj.MPD;
    if (!mpd) {
      logger.error("[ManifestParser] No MPD element found in manifest");
      return [];
    }

    const period = mpd.Period;
    if (!period) {
      logger.error("[ManifestParser] No Period element found in MPD");
      return [];
    }

    // BaseURL might be at MPD or Period level
    let mpdBaseUrl = "";
    if (mpd.BaseURL) {
      mpdBaseUrl =
        typeof mpd.BaseURL === "string" ? mpd.BaseURL : mpd.BaseURL["#text"];
    } else if (manifestUrl) {
      // Fallback to manifest URL base
      // YouTube manifest.dash URLs are like .../manifest.dash
      // Segments are usually .../sq/1
      // So we can remove "manifest.dash" or just append to the base path.
      // But usually YouTube DASH segments are siblings to manifest.dash in the URL structure logic.
      // Actually, for YouTube, the BaseURL is effectively the same as manifest URL minus the command.
      // Safest bet: If manifestUrl ends with 'manifest.dash', remove it.
      mpdBaseUrl = manifestUrl.replace("manifest.dash", "");
      if (!mpdBaseUrl.endsWith("/")) mpdBaseUrl += "/"; // ensure trailing slash if we are appending relative paths
    }

    // Handle AdaptationSets (Video/Audio)
    if (!period.AdaptationSet) {
      logger.error("[ManifestParser] No AdaptationSet found in Period");
      return [];
    }

    const adaptationSets = Array.isArray(period.AdaptationSet)
      ? period.AdaptationSet
      : [period.AdaptationSet];
    const streams: DashStream[] = [];

    logger.debug(
      `[ManifestParser] Found ${adaptationSets.length} AdaptationSet(s)`,
    );

    // Check for SegmentList at Period level (YouTube live streams)
    const periodSegmentList = period.SegmentList;
    let periodStartNumber = 0;
    let periodTimescale = 1000;
    const periodSegments: DashSegment[] = [];

    if (periodSegmentList) {
      periodStartNumber = parseInt(periodSegmentList["@_startNumber"] || "0", 10);
      periodTimescale = parseInt(periodSegmentList["@_timescale"] || "1000", 10);

      // Parse SegmentTimeline from Period-level SegmentList
      const timeline = periodSegmentList.SegmentTimeline;
      if (timeline && timeline.S) {
        const sList = Array.isArray(timeline.S) ? timeline.S : [timeline.S];
        for (const s of sList) {
          periodSegments.push({
            t: s["@_t"] ? parseInt(s["@_t"], 10) : undefined,
            d: parseInt(s["@_d"], 10),
            r: s["@_r"] ? parseInt(s["@_r"], 10) : undefined,
          });
        }
      }
      logger.debug(
        `[ManifestParser] Found Period-level SegmentList: startNumber=${periodStartNumber}, timescale=${periodTimescale}, segments=${periodSegments.length}`,
      );
    }

    for (const adaptSet of adaptationSets) {
      if (!adaptSet.Representation) {
        logger.debug(
          `[ManifestParser] AdaptationSet has no Representation, skipping`,
        );
        continue;
      }

      const representations = Array.isArray(adaptSet.Representation)
        ? adaptSet.Representation
        : [adaptSet.Representation];

      logger.debug(
        `[ManifestParser] AdaptationSet mimeType=${adaptSet["@_mimeType"]} has ${representations.length} Representation(s)`,
      );

      for (const rep of representations) {
        // Check for BaseURL at AdaptationSet or Representation level
        let repBaseUrl = mpdBaseUrl;
        if (period.BaseURL)
          repBaseUrl =
            typeof period.BaseURL === "string"
              ? period.BaseURL
              : period.BaseURL["#text"];
        if (adaptSet.BaseURL)
          repBaseUrl =
            typeof adaptSet.BaseURL === "string"
              ? adaptSet.BaseURL
              : adaptSet.BaseURL["#text"];
        if (rep.BaseURL)
          repBaseUrl =
            typeof rep.BaseURL === "string"
              ? rep.BaseURL
              : rep.BaseURL["#text"];

        const mimeType = rep["@_mimeType"] || adaptSet["@_mimeType"];
        const codecs = rep["@_codecs"] || adaptSet["@_codecs"];
        const width = rep["@_width"];
        const height = rep["@_height"];
        const bandwidth = parseInt(rep["@_bandwidth"], 10);
        const itag = parseInt(rep["@_id"], 10);

        // SegmentTemplate might be in Representation or AdaptationSet
        const segTemplate = rep.SegmentTemplate || adaptSet.SegmentTemplate;

        // For YouTube live streams: use Period-level SegmentList with Representation BaseURL
        if (!segTemplate && periodSegmentList && repBaseUrl) {
          // YouTube live format: BaseURL contains the full segment base URL
          // Log the full BaseURL to understand the format
          logger.debug(
            `[ManifestParser] Raw BaseURL for itag=${itag}: ${repBaseUrl}`,
          );

          let mediaTemplate = repBaseUrl;

          // Check if BaseURL already contains sq/ pattern anywhere
          if (mediaTemplate.includes("/sq/")) {
            // Replace the segment number with template placeholder
            mediaTemplate = mediaTemplate.replace(/\/sq\/\d+/, "/sq/$Number$");
          } else {
            // BaseURL is the base path, append sq/$Number$
            if (!mediaTemplate.endsWith("/")) {
              mediaTemplate += "/";
            }
            mediaTemplate += "sq/$Number$";
          }

          streams.push({
            itag,
            mimeType,
            codecs,
            width,
            height,
            bandwidth,
            baseUrl: mediaTemplate,
            startNumber: periodStartNumber,
            timescale: periodTimescale,
            segments: periodSegments,
            initialization: undefined, // YouTube live doesn't use separate init segments
          });
          logger.debug(
            `[ManifestParser] Final template for itag=${itag}: ${mediaTemplate}`,
          );
          continue;
        }

        if (!segTemplate) {
          logger.debug(
            `[ManifestParser] Representation itag=${itag} has no SegmentTemplate and no Period SegmentList, skipping`,
          );
          continue;
        }

        const startNumber = parseInt(segTemplate["@_startNumber"] || "0", 10);
        const timescale = parseInt(segTemplate["@_timescale"] || "1000", 10);
        let initialization = segTemplate["@_initialization"];
        let mediaTemplate = segTemplate["@_media"]; // "sq/$Number$" or similar

        // Resolve relative URLs
        if (repBaseUrl) {
          if (mediaTemplate && !mediaTemplate.startsWith("http")) {
            mediaTemplate = repBaseUrl + mediaTemplate;
          }
          if (initialization && !initialization.startsWith("http")) {
            initialization = repBaseUrl + initialization;
          }
        }

        // Parse SegmentTimeline
        const timeline = segTemplate.SegmentTimeline;
        const segments: DashSegment[] = [];
        if (timeline && timeline.S) {
          const sList = Array.isArray(timeline.S) ? timeline.S : [timeline.S];
          for (const s of sList) {
            segments.push({
              t: s["@_t"] ? parseInt(s["@_t"], 10) : undefined,
              d: parseInt(s["@_d"], 10),
              r: s["@_r"] ? parseInt(s["@_r"], 10) : undefined,
            });
          }
        }

        streams.push({
          itag,
          mimeType,
          codecs,
          width,
          height,
          bandwidth,
          baseUrl: mediaTemplate, // Store the template
          startNumber,
          timescale,
          segments,
          initialization,
        });
      }
    }

    return streams;
  }

  static parseHls(
    m3u8Content: string,
    baseUrl: string,
  ): {
    type: "master" | "media";
    variants?: ParsedHlsVariant[];
    segments?: ParsedHlsSegment[];
    targetDuration?: number;
    mediaSequence?: number;
    endList?: boolean;
    discontinuitySequence?: number;
  } {
    const lines = m3u8Content.split("\n");

    let isMaster = false;
    const variants: ParsedHlsVariant[] = [];
    const segments: ParsedHlsSegment[] = [];
    let targetDuration = 0;
    let mediaSequence = 0;
    let endList = false;
    let discontinuitySequence = 0;
    let currentDiscontinuity = 0;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i].trim();
      if (!line) continue;

      if (line.startsWith("#EXT-X-STREAM-INF:")) {
        isMaster = true;
        const bandwidthMatch = line.match(/BANDWIDTH=(\d+)/);
        const resMatch = line.match(/RESOLUTION=(\d+)x(\d+)/);
        const codecsMatch = line.match(/CODECS="([^"]+)"/);

        let url = lines[i + 1]?.trim();
        if (url && !url.startsWith("#")) {
          if (!url.startsWith("http")) {
            url = new URL(url, baseUrl).toString();
          }

          variants.push({
            bandwidth: bandwidthMatch ? parseInt(bandwidthMatch[1], 10) : 0,
            width: resMatch ? parseInt(resMatch[1], 10) : 0,
            height: resMatch ? parseInt(resMatch[2], 10) : 0,
            codecs: codecsMatch ? codecsMatch[1] : "",
            url: url,
          });
          i++;
        }
      } else if (line.startsWith("#EXT-X-TARGETDURATION:")) {
        targetDuration = parseFloat(line.split(":")[1]);
      } else if (line.startsWith("#EXT-X-MEDIA-SEQUENCE:")) {
        mediaSequence = parseInt(line.split(":")[1], 10);
      } else if (line.startsWith("#EXT-X-DISCONTINUITY-SEQUENCE:")) {
        discontinuitySequence = parseInt(line.split(":")[1], 10);
      } else if (line === "#EXT-X-DISCONTINUITY") {
        currentDiscontinuity++;
      } else if (line === "#EXT-X-ENDLIST") {
        endList = true;
      } else if (line.startsWith("#EXTINF:")) {
        const duration = parseFloat(line.split(":")[1].split(",")[0]);
        let url = lines[i + 1]?.trim();
        if (url && !url.startsWith("#")) {
          if (!url.startsWith("http")) {
            url = new URL(url, baseUrl).toString();
          }
          segments.push({
            duration: duration,
            url: url,
            discontinuity: currentDiscontinuity,
          });
          i++;
        }
      }
    }

    if (isMaster) {
      return { type: "master", variants };
    } else {
      return {
        type: "media",
        segments,
        targetDuration,
        mediaSequence,
        endList,
        discontinuitySequence,
      };
    }
  }

  /**
   * Calculate which segments to download for a given time range.
   *
   * For DASH streams with SegmentTimeline, maps start/end timestamps (in seconds)
   * to segment indices. Returns the segment range and trim offsets for the muxer.
   *
   * @param stream - The DASH stream with segment timeline info
   * @param startTimeSec - Start time in seconds (0 = beginning)
   * @param endTimeSec - End time in seconds (undefined = end of stream)
   * @returns Segment range and trim info, or null if timeline is unavailable
   */
  static calculateSegmentRange(
    stream: DashStream,
    startTimeSec?: number,
    endTimeSec?: number,
  ): SegmentRange | null {
    const logger = Logger.getInstance();

    // YouTube live streams typically use ~2s segments without SegmentTimeline
    // For these, we estimate based on a fixed segment duration
    const segments = stream.segments;
    const timescale = stream.timescale || 1000;

    // Calculate segment durations in seconds
    // Expand the segment timeline (handling repeat counts)
    const expandedDurations: number[] = [];
    if (segments.length > 0) {
      for (const seg of segments) {
        const durationSec = seg.d / timescale;
        const repeatCount = (seg.r ?? 0) + 1; // r=0 means 1 occurrence, r=2 means 3
        for (let i = 0; i < repeatCount; i++) {
          expandedDurations.push(durationSec);
        }
      }
    }

    // If no segment timeline is available, estimate with default segment duration
    // YouTube live typically uses ~2 second segments
    const DEFAULT_SEGMENT_DURATION = 2.0;
    const useEstimate = expandedDurations.length === 0;

    if (useEstimate) {
      logger.debug(
        "[ManifestParser] No SegmentTimeline available, using estimated segment duration",
      );
    }

    const getSegmentDuration = (idx: number): number => {
      if (useEstimate) return DEFAULT_SEGMENT_DURATION;
      return idx < expandedDurations.length
        ? expandedDurations[idx]
        : expandedDurations[expandedDurations.length - 1] || DEFAULT_SEGMENT_DURATION;
    };

    // Walk through segments to find start and end indices
    let cumulativeTime = 0;
    let startSegment = stream.startNumber;
    let trimStartOffset = 0;

    if (startTimeSec != null && startTimeSec > 0) {
      let segIdx = 0;
      while (true) {
        const segDuration = getSegmentDuration(segIdx);
        if (cumulativeTime + segDuration > startTimeSec) {
          // Start time falls within this segment
          startSegment = stream.startNumber + segIdx;
          trimStartOffset = startTimeSec - cumulativeTime;
          break;
        }
        cumulativeTime += segDuration;
        segIdx++;

        // Safety: don't scan beyond a reasonable limit
        if (segIdx > 100_000) {
          startSegment = stream.startNumber + segIdx;
          trimStartOffset = 0;
          break;
        }
      }
    }

    let endSegment = -1; // -1 = download to end
    let totalRequestedDuration = 0;

    if (endTimeSec != null && endTimeSec > 0) {
      cumulativeTime = 0;
      let segIdx = 0;
      while (true) {
        const segDuration = getSegmentDuration(segIdx);
        cumulativeTime += segDuration;
        if (cumulativeTime >= endTimeSec) {
          endSegment = stream.startNumber + segIdx;
          break;
        }
        segIdx++;

        // Safety: don't scan beyond a reasonable limit
        if (segIdx > 100_000) {
          endSegment = stream.startNumber + segIdx;
          break;
        }
      }

      totalRequestedDuration = endTimeSec - (startTimeSec || 0);
    }

    logger.debug(
      `[ManifestParser] Segment range: ${startSegment}-${endSegment === -1 ? "end" : endSegment}, ` +
        `trimStart=${trimStartOffset.toFixed(2)}s, duration=${totalRequestedDuration > 0 ? totalRequestedDuration.toFixed(2) + "s" : "full"}`,
    );

    return {
      startSegment,
      endSegment,
      trimStartOffset,
      trimDuration: totalRequestedDuration,
    };
  }
}

File: internal/engine/manifest.go
package engine

import (
	"encoding/xml"
	"fmt"
	"math"
	"net/url"
	"regexp"
	"strconv"
	"strings"
)

const defaultSegmentDuration = 2.0

// DashSegment represents a segment in a DASH timeline.
type DashSegment struct {
	T int64 // Timestamp (optional, 0 if unset)
	D int64 // Duration in timescale units
	R int   // Repeat count (0 means 1 occurrence, 1 means 2, etc.)
}

// DashStream represents a parsed DASH adaptation/representation.
type DashStream struct {
	Itag           int
	MimeType       string
	Codecs         string
	Width          int
	Height         int
	FPS            int           // From frameRate attribute (0 if not present)
	Bandwidth      int
	BaseURL        string        // Template URL with $Number$ placeholder
	StartNumber    int           // First segment sequence number
	Timescale      int           // Time unit divisor
	Segments       []DashSegment // Segment timeline
	Initialization string        // Init segment URL
}

// SegmentRange maps a time range to segment indices.
type SegmentRange struct {
	StartSegment    int     // Absolute segment index
	EndSegment      int     // Inclusive, -1 means to end
	TrimStartOffset float64 // Seconds into first segment
	TrimDuration    float64 // Total duration
}

// HlsVariant represents a variant in an HLS master playlist.
type HlsVariant struct {
	Bandwidth int
	Width     int
	Height    int
	FPS       int // From FRAME-RATE attribute (0 if not present)
	Codecs    string
	URL       string
	Name      string
	IsSource  bool
}

// HlsSegment represents a segment in an HLS media playlist.
type HlsSegment struct {
	Duration      float64
	URL           string
	Discontinuity int
}

// HlsPlaylist represents a parsed HLS media playlist.
type HlsPlaylist struct {
	Segments              []HlsSegment
	TargetDuration        float64
	MediaSequence         int
	DiscontinuitySequence int
	EndList               bool
}

// HlsParseResult contains either master variants or media segments.
type HlsParseResult struct {
	IsMaster bool
	Variants []HlsVariant
	Playlist *HlsPlaylist
}

// --- DASH Parsing ---

// XML structures for DASH MPD parsing.
type mpdXML struct {
	XMLName xml.Name    `xml:"MPD"`
	BaseURL string      `xml:"BaseURL"`
	Periods []periodXML `xml:"Period"`
}

type periodXML struct {
	BaseURL        string             `xml:"BaseURL"`
	AdaptationSets []adaptationSetXML `xml:"AdaptationSet"`
	SegmentList    *segmentListXML    `xml:"SegmentList"`
}

type adaptationSetXML struct {
	MimeType        string              `xml:"mimeType,attr"`
	Codecs          string              `xml:"codecs,attr"`
	BaseURL         string              `xml:"BaseURL"`
	Representations []representationXML `xml:"Representation"`
	SegmentTemplate *segmentTemplateXML `xml:"SegmentTemplate"`
	SegmentList     *segmentListXML     `xml:"SegmentList"`
}

type representationXML struct {
	ID              string              `xml:"id,attr"`
	Bandwidth       int                 `xml:"bandwidth,attr"`
	Width           int                 `xml:"width,attr"`
	Height          int                 `xml:"height,attr"`
	FrameRate       string              `xml:"frameRate,attr"` // e.g. "30", "60", "30/1"
	Codecs          string              `xml:"codecs,attr"`
	MimeType        string              `xml:"mimeType,attr"`
	BaseURL         string              `xml:"BaseURL"`
	SegmentTemplate *segmentTemplateXML `xml:"SegmentTemplate"`
	SegmentList     *segmentListXML     `xml:"SegmentList"`
}

type segmentTemplateXML struct {
	Media          string           `xml:"media,attr"`
	Initialization string           `xml:"initialization,attr"`
	StartNumber    int              `xml:"startNumber,attr"`
	Timescale      int              `xml:"timescale,attr"`
	Timeline       *segmentTimeline `xml:"SegmentTimeline"`
}

type segmentListXML struct {
	StartNumber    int              `xml:"startNumber,attr"`
	Timescale      int              `xml:"timescale,attr"`
	Initialization *initXML         `xml:"Initialization"`
	Timeline       *segmentTimeline `xml:"SegmentTimeline"`
}

type segmentTimeline struct {
	Segments []segmentTimelineS `xml:"S"`
}

type segmentTimelineS struct {
	T int64 `xml:"t,attr"`
	D int64 `xml:"d,attr"`
	R int   `xml:"r,attr"`
}

type initXML struct {
	SourceURL string `xml:"sourceURL,attr"`
}

var sqPattern = regexp.MustCompile(`/sq/\d+`)

// ParseDash parses a DASH MPD manifest and returns available streams.
func ParseDash(xmlContent string, manifestURL string) ([]DashStream, error) {
	var mpd mpdXML
	if err := xml.Unmarshal([]byte(xmlContent), &mpd); err != nil {
		return nil, fmt.Errorf("parse MPD XML: %w", err)
	}

	if len(mpd.Periods) == 0 {
		return nil, nil
	}

	mpdBaseURL := mpd.BaseURL
	if mpdBaseURL == "" && manifestURL != "" {
		if u, err := url.Parse(manifestURL); err == nil {
			if idx := strings.LastIndex(u.Path, "/"); idx >= 0 {
				u.Path = u.Path[:idx+1]
			}
			// Preserve query string — YouTube uses it for token auth
			mpdBaseURL = u.String()
		}
	}

	var streams []DashStream

	for _, period := range mpd.Periods {
		periodBase := mpdBaseURL
		if period.BaseURL != "" {
			if strings.HasPrefix(period.BaseURL, "http") {
				periodBase = period.BaseURL
			} else if mpdBaseURL != "" {
				periodBase = resolveURL(mpdBaseURL, period.BaseURL)
			}
		}

		for _, as := range period.AdaptationSets {
			for _, rep := range as.Representations {
				stream := parseDashRepresentation(rep, as, period, periodBase)
				if stream != nil {
					streams = append(streams, *stream)
				}
			}
		}
	}

	return streams, nil
}

func parseDashRepresentation(rep representationXML, as adaptationSetXML, period periodXML, baseURL string) *DashStream {
	stream := &DashStream{
		Bandwidth: rep.Bandwidth,
		Width:     rep.Width,
		Height:    rep.Height,
		FPS:       parseFrameRate(rep.FrameRate),
	}

	// Parse itag from ID
	if rep.ID != "" {
		var err error
		stream.Itag, err = strconv.Atoi(rep.ID)
		if err != nil {
			// Non-numeric representation ID — itag stays 0
			stream.Itag = 0
		}
	}

	// MimeType: representation > adaptation set
	stream.MimeType = rep.MimeType
	if stream.MimeType == "" {
		stream.MimeType = as.MimeType
	}

	// Codecs: representation > adaptation set
	stream.Codecs = rep.Codecs
	if stream.Codecs == "" {
		stream.Codecs = as.Codecs
	}

	// BaseURL resolution chain: Representation > AdaptationSet > Period > MPD
	repBase := baseURL
	if as.BaseURL != "" {
		if strings.HasPrefix(as.BaseURL, "http") {
			repBase = as.BaseURL
		} else if baseURL != "" {
			repBase = resolveURL(baseURL, as.BaseURL)
		}
	}
	if rep.BaseURL != "" {
		if strings.HasPrefix(rep.BaseURL, "http") {
			repBase = rep.BaseURL
		} else if repBase != "" {
			repBase = resolveURL(repBase, rep.BaseURL)
		} else {
			repBase = rep.BaseURL
		}
	}

	// YouTube live: convert /sq/N to /sq/$Number$
	if sqPattern.MatchString(repBase) {
		repBase = sqPattern.ReplaceAllString(repBase, "/sq/$Number$")
	}
	stream.BaseURL = repBase

	// Segment template: representation > adaptation set > period
	tmpl := rep.SegmentTemplate
	if tmpl == nil {
		tmpl = as.SegmentTemplate
	}

	segList := rep.SegmentList
	if segList == nil {
		segList = as.SegmentList
	}
	if segList == nil {
		segList = period.SegmentList
	}

	if tmpl != nil {
		stream.StartNumber = tmpl.StartNumber
		stream.Timescale = tmpl.Timescale
		if stream.Timescale == 0 {
			stream.Timescale = 1000
		}
		if tmpl.Media != "" {
			stream.BaseURL = resolveURL(repBase, tmpl.Media)
		}
		if tmpl.Initialization != "" {
			stream.Initialization = resolveURL(repBase, tmpl.Initialization)
		}
		if tmpl.Timeline != nil {
			stream.Segments = parseTimeline(tmpl.Timeline)
		}
	} else if segList != nil {
		stream.StartNumber = segList.StartNumber
		stream.Timescale = segList.Timescale
		if stream.Timescale == 0 {
			stream.Timescale = 1000
		}
		if segList.Initialization != nil && segList.Initialization.SourceURL != "" {
			stream.Initialization = resolveURL(repBase, segList.Initialization.SourceURL)
		}
		if segList.Timeline != nil {
			stream.Segments = parseTimeline(segList.Timeline)
		}
	}

	return stream
}

func parseTimeline(tl *segmentTimeline) []DashSegment {
	segments := make([]DashSegment, 0, len(tl.Segments))
	for _, s := range tl.Segments {
		segments = append(segments, DashSegment(s))
	}
	return segments
}

// SegmentURL builds the URL for a specific segment number from a template.
func SegmentURL(template string, seqNum int) string {
	return strings.ReplaceAll(template, "$Number$", strconv.Itoa(seqNum))
}

// CalculateSegmentRange maps a time range to segment indices in a DASH stream.
func CalculateSegmentRange(stream *DashStream, startTimeSec, endTimeSec float64) *SegmentRange {
	timescale := float64(stream.Timescale)
	if timescale == 0 {
		timescale = 1000 // Default fallback (matches TS)
	}

	// Expand segments with repeat counts
	type expandedSeg struct {
		duration float64 // seconds
	}
	var expanded []expandedSeg

	if len(stream.Segments) > 0 {
		for _, seg := range stream.Segments {
			dur := float64(seg.D) / timescale
			count := seg.R + 1 // R=0 means 1 occurrence
			for range count {
				expanded = append(expanded, expandedSeg{duration: dur})
				if len(expanded) > 100000 {
					return nil // Safety limit
				}
			}
		}
	}

	// If no segment timeline, use estimated 2s segment duration (YouTube live typical)
	const defaultSegDuration = 2.0
	useEstimate := len(expanded) == 0

	getSegDuration := func(idx int) float64 {
		if useEstimate {
			return defaultSegDuration
		}
		if idx < len(expanded) {
			return expanded[idx].duration
		}
		if len(expanded) > 0 {
			return expanded[len(expanded)-1].duration
		}
		return defaultSegDuration
	}

	result := &SegmentRange{
		StartSegment: stream.StartNumber,
		EndSegment:   -1,
	}

	// A8: Use stream.StartNumber to convert array index to actual segment number
	baseNum := stream.StartNumber

	// Walk through segments to find start index
	if startTimeSec > 0 {
		accumSec := 0.0
		segIdx := 0
		for {
			segDur := getSegDuration(segIdx)
			if accumSec+segDur > startTimeSec {
				result.StartSegment = baseNum + segIdx
				result.TrimStartOffset = startTimeSec - accumSec
				break
			}
			accumSec += segDur
			segIdx++
			if segIdx > 100000 {
				result.StartSegment = baseNum + segIdx
				result.TrimStartOffset = 0
				break
			}
		}
	}

	// Walk through segments to find end index
	if endTimeSec > 0 {
		accumSec := 0.0
		segIdx := 0
		for {
			segDur := getSegDuration(segIdx)
			segEnd := accumSec + segDur
			if segEnd >= endTimeSec {
				result.EndSegment = baseNum + segIdx
				result.TrimDuration = endTimeSec - startTimeSec
				break
			}
			accumSec = segEnd
			segIdx++
			if segIdx > 100000 {
				break
			}
		}
	}

	return result
}

// --- HLS Parsing ---

// ParseHls parses an HLS M3U8 playlist (master or media).
func ParseHls(m3u8Content string, baseURL string) *HlsParseResult {
	lines := strings.Split(strings.TrimSpace(m3u8Content), "\n")
	if len(lines) == 0 {
		return nil
	}

	// Check if master or media
	isMaster := false
	for _, line := range lines {
		if strings.HasPrefix(strings.TrimSpace(line), "#EXT-X-STREAM-INF") {
			isMaster = true
			break
		}
	}

	if isMaster {
		return parseMasterPlaylist(lines, baseURL)
	}
	return parseMediaPlaylist(lines, baseURL)
}

func parseMasterPlaylist(lines []string, baseURL string) *HlsParseResult {
	result := &HlsParseResult{IsMaster: true}

	for i := 0; i < len(lines); i++ {
		line := strings.TrimSpace(lines[i])
		if !strings.HasPrefix(line, "#EXT-X-STREAM-INF:") {
			continue
		}

		attrs := line[len("#EXT-X-STREAM-INF:"):]
		variant := HlsVariant{}

		// Parse attributes
		for _, attr := range parseAttributes(attrs) {
			switch attr.key {
			case "BANDWIDTH":
				variant.Bandwidth, _ = strconv.Atoi(attr.value)
			case "RESOLUTION":
				if w, h, ok := strings.Cut(attr.value, "x"); ok {
					variant.Width, _ = strconv.Atoi(w)
					variant.Height, _ = strconv.Atoi(h)
				}
			case "FRAME-RATE":
				if v, err := strconv.ParseFloat(attr.value, 64); err == nil {
					variant.FPS = int(math.Round(v))
				}
			case "CODECS":
				variant.Codecs = strings.Trim(attr.value, "\"")
			case "VIDEO":
				name := strings.Trim(attr.value, "\"")
				variant.Name = name
				variant.IsSource = name == "chunked" || strings.Contains(strings.ToLower(name), "source")
			}
		}

		// Next non-comment line is the URL
		for i++; i < len(lines); i++ {
			nextLine := strings.TrimSpace(lines[i])
			if nextLine != "" && !strings.HasPrefix(nextLine, "#") {
				variant.URL = resolveURL(baseURL, nextLine)
				break
			}
		}

		result.Variants = append(result.Variants, variant)
	}

	return result
}

func parseMediaPlaylist(lines []string, baseURL string) *HlsParseResult {
	playlist := &HlsPlaylist{}
	discontinuity := 0
	var currentDuration float64

	for _, rawLine := range lines {
		line := strings.TrimSpace(rawLine)

		switch {
		case strings.HasPrefix(line, "#EXT-X-TARGETDURATION:"):
			playlist.TargetDuration, _ = strconv.ParseFloat(line[len("#EXT-X-TARGETDURATION:"):], 64)

		case strings.HasPrefix(line, "#EXT-X-MEDIA-SEQUENCE:"):
			playlist.MediaSequence, _ = strconv.Atoi(line[len("#EXT-X-MEDIA-SEQUENCE:"):])

		case strings.HasPrefix(line, "#EXT-X-DISCONTINUITY-SEQUENCE:"):
			playlist.DiscontinuitySequence, _ = strconv.Atoi(line[len("#EXT-X-DISCONTINUITY-SEQUENCE:"):])

		case strings.HasPrefix(line, "#EXTINF:"):
			durationStr := line[len("#EXTINF:"):]
			durationStr, _, _ = strings.Cut(durationStr, ",")
			currentDuration, _ = strconv.ParseFloat(durationStr, 64)

		case line == "#EXT-X-DISCONTINUITY":
			discontinuity++

		case line == "#EXT-X-ENDLIST":
			playlist.EndList = true

		case line != "" && !strings.HasPrefix(line, "#"):
			// Segment URL
			segment := HlsSegment{
				Duration:      currentDuration,
				URL:           resolveURL(baseURL, line),
				Discontinuity: discontinuity,
			}
			playlist.Segments = append(playlist.Segments, segment)
			currentDuration = 0
		}
	}

	return &HlsParseResult{Playlist: playlist}
}

// --- Utility ---

type attribute struct {
	key   string
	value string
}

func parseAttributes(s string) []attribute {
	var attrs []attribute
	for len(s) > 0 {
		s = strings.TrimSpace(s)
		key, rest, ok := strings.Cut(s, "=")
		if !ok {
			break
		}
		s = rest

		var value string
		if len(s) > 0 && s[0] == '"' {
			// Quoted value
			if inner, after, ok := strings.Cut(s[1:], "\""); ok {
				value = inner
				s = after
			} else {
				value = s[1:]
				s = ""
			}
		} else {
			value, s, _ = strings.Cut(s, ",")
		}

		if len(s) > 0 && s[0] == ',' {
			s = s[1:]
		}
		attrs = append(attrs, attribute{key: key, value: value})
	}
	return attrs
}

func resolveURL(base, ref string) string {
	if strings.HasPrefix(ref, "http://") || strings.HasPrefix(ref, "https://") {
		return ref
	}
	if base == "" {
		return ref
	}
	baseU, err := url.Parse(base)
	if err != nil {
		return ref
	}
	refU, err := url.Parse(ref)
	if err != nil {
		return ref
	}
	return baseU.ResolveReference(refU).String()
}

// TotalDurationSec returns the total duration of a DASH stream in seconds.
func TotalDurationSec(stream *DashStream) float64 {
	if stream.Timescale == 0 || len(stream.Segments) == 0 {
		return 0
	}
	ts := float64(stream.Timescale)
	total := 0.0
	for _, seg := range stream.Segments {
		count := seg.R + 1
		total += float64(seg.D) / ts * float64(count)
	}
	return total
}

// TotalSegmentCount returns the total number of segments, expanding repeats.
func TotalSegmentCount(stream *DashStream) int {
	total := 0
	for _, seg := range stream.Segments {
		total += seg.R + 1
	}
	return total
}

// SegmentDurationSec returns the average segment duration in seconds.
func SegmentDurationSec(stream *DashStream) float64 {
	count := TotalSegmentCount(stream)
	if count == 0 {
		return defaultSegmentDuration
	}
	return TotalDurationSec(stream) / float64(count)
}

// EstimateSegmentCount estimates the number of segments for a given duration.
func EstimateSegmentCount(durationSec float64) int {
	return int(math.Ceil(durationSec / defaultSegmentDuration))
}

// parseFrameRate parses a DASH frameRate attribute value.
// YouTube uses plain integers ("30", "60") but the spec also allows
// fractional notation ("30000/1001"). Returns 0 if empty or unparseable.
func parseFrameRate(s string) int {
	if s == "" {
		return 0
	}
	// Try fractional "num/den"
	if idx := strings.Index(s, "/"); idx > 0 {
		num, err1 := strconv.Atoi(s[:idx])
		den, err2 := strconv.Atoi(s[idx+1:])
		if err1 == nil && err2 == nil && den > 0 {
			return int(math.Round(float64(num) / float64(den)))
		}
	}
	// Plain integer
	if v, err := strconv.Atoi(s); err == nil {
		return v
	}
	// Float string (rare)
	if v, err := strconv.ParseFloat(s, 64); err == nil {
		return int(math.Round(v))
	}
	return 0
}
